astropy/utils/metadata.py

```diff
@@ -422,30 +423,79 @@ class MetaData:
         Default is ``True``.
 
         .. versionadded:: 1.2
+
+    Examples
+    --------
+    ``MetaData`` can be used as a descriptor to define a ``meta`` attribute`.
+
+        >>> class Foo:
+        ...     meta = MetaData()
+        ...     def __init__(self, meta):
+        ...         self.meta = meta
+
+    If ``Foo`` is not a dataclass, then ``Foo.meta`` is returns the descriptor.
+
+        >>> print(Foo.meta)
+        <astropy.utils.metadata.MetaData object at ...>
+
+    ``Foo`` can be instantiated with a ``meta`` argument.
+
+        >>> foo = Foo(meta={'a': 1, 'b': 2})
+        >>> foo.meta
+        {'a': 1, 'b': 2}
+
+    We can also use ``MetaData`` as a field in a dataclass.
+
+        >>> from dataclasses import dataclass
+        >>> from astropy.utils.metadata import MetaData
+        >>> @dataclass
+        ... class Bar:
+        ...     meta: dict = MetaData()
+
+    On a dataclass, ``Bar.meta`` returns `None`.
+
+        >>> print(Bar.meta)
+        None
+
+    ``Bar`` can be instantiated with a ``meta`` argument.
+
+        >>> bar = Bar(meta={'a': 1, 'b': 2})
+        >>> bar.meta
+        {'a': 1, 'b': 2}
     """
 
     def __init__(self, doc="", copy=True):
         self.__doc__ = doc
         self.copy = copy
 
     def __get__(self, instance, owner):
+        # class attribute access
         if instance is None:
-            return self
+            return None if is_dataclass(owner) else self
+        # instance attribute access
         if not hasattr(instance, "_meta"):
             instance._meta = OrderedDict()
         return instance._meta
 
     def __set__(self, instance, value):
+        # The 'default' value (as set on a dataclass) is `None`, but we want to set it
+        # to an empty `OrderedDict` if it is `None` so that we can always assume it is a
+        # `Mapping` and not have to check for `None` everywhere.
         if value is None:
             instance._meta = OrderedDict()
+        # The default value on a dataclass is determined before the dataclass is
```

Comment:
This `if` block is really showing that it's hard to gently transition `MetaData` to work with dataclasses. With this PR `dataclasses.fields(X)` actually returns incorrect information, since the default value is intended to be `None`, not `self`.
A followup PR will be to remove `self` as an output, leaving just `None`. This prevents easy access to the descriptor from the class, but that's fine since it's no different conceptually than accessing the wrapped function of a `property` and we don't do that like we don't really need to access the descriptor on the class. And if it's really necessary, descriptors can be gotten by `vars(cls)["<descriptor attribute name>"]`.

Reply:
I'm fine with removing ever returning `self`

Reply:
Then I'll make this an API changelog!

Reply:
Done!

Reply:
Not needed and now removed.

Reply:
✅